platform/src/server/http/http_server.ts

```diff
@@ -1,51 +1,82 @@
-import express from 'express';
-import http from 'http';
-import https from 'https';
-import { Server } from 'net';
-import { promisify } from 'util';
+import { Request, ResponseToolkit, Server, ServerOptions } from 'hapi';
+import { ServerOptions as TLSOptions} from 'https';
 import { readFileSync } from 'fs';
 
 import { HttpConfig } from './http_config';
 import { Env } from '../../config';
 import { Logger } from '../../logging';
-import { Router } from './Router';
+import { Router } from './router';
+import { modifyUrl } from '../../lib/utils/url';
 
 export class HttpServer {
-  private readonly app: express.Application;
   private server?: Server;
+  private registeredRouters: Set<Router> = new Set();
 
   constructor(private readonly log: Logger, private readonly env: Env) {
-    this.app = express();
   }
 
   isListening() {
-    return this.server !== undefined && (this.server as any).listening;
+    return this.server !== undefined && this.server.listener.listening;
   }
 
   registerRouter(router: Router) {
-    this.app.use(router.path, router.router);
+    if (this.isListening()) {
+      throw new Error('Routers can be registered only when HTTP server is stopped.')
+    }
+
+    this.registeredRouters.add(router);
   }
 
   async start(config: HttpConfig) {
-    const server = this.initializeServer(config);
-
-    this.server = server;
+    this.server = this.initializeServer(config);
+
+    this.setupBasePathRewrite(this.server, config);
+
+    for (const router of this.registeredRouters) {
+      for (const route of router.getRoutes()) {
+        this.server.route({
+          method: route.method,
+          path: `${router.path}${route.path}`,
+          handler: route.handler,
+        });
+      }
+    }
 
     const legacyKbnServer = this.env.getLegacyKbnServer();
     if (legacyKbnServer !== undefined) {
-      legacyKbnServer.newPlatformProxyListener.bind(server);
+      legacyKbnServer.newPlatformProxyListener.bind(this.server.listener);
 
       // We register Kibana proxy middleware right before we start server to allow
-      // all new platform plugins register their endpoints, so that kbnServer
+      // all new platform plugins register their routes, so that kbnServer
       // handles only requests that aren't handled by the new platform.
-      this.app.use((req, res) =>
-        legacyKbnServer.newPlatformProxyListener.proxy(req, res)
-      );
+      this.server.route({
+        method: '*',
+        path: '/{p*}',
+        options: {
+          payload: {
+            output: 'stream',
+            parse: false,
+            timeout: false,
+          }
+        },
+        handler: ({ raw: { req, res }}, h) => {
+          legacyKbnServer.newPlatformProxyListener.proxy(req, res);
+          return h.abandon;
```

Comment:
nit: One letter variable names make it more difficult to understand the code. I think short variable names for super common things like `req`, `res`, and `err` are fine, but it would be helpful if we used more expressive names otherwise.

Reply:
Yeah, agree, used the convention established in Hapi docs by default, but will rename it into something that's clearer.

Reply:
:heavy_check_mark: